include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h

```diff
@@ -1411,7 +1411,7 @@ __pattern_partial_sort_copy(_ExecutionPolicy&& __exec, _InIterator __first, _InI
     {
         // If our output buffer is larger than the input buffer, simply copy elements to the output and use
         // full sort on them.
-        auto __out_end = __pattern_walk2</*_IsSync=*/::std::false_type>(
+        auto __out_end = __pattern_walk2(
             __par_backend_hetero::make_wrapped_policy<__initial_copy_1>(__exec), __first, __last, __out_first,
```

Comment:
This `__pattern_walk2` and the subsequent `__pattern_sort` do have a dependency with `__out_first`, where `__pattern_walk2` needs to be finished before `__pattern_sort` can begin.   

My first thought is that the accessors involved here provide the implicit data dependency to order these kernels appropriately and make this work as a pipeline (as described in the `_IsSync` comment in the definition of `__pattern_walk2`).

However, with USM pointers as input data, I believe we lose these implicit dependencies and the ordering of kernels.  Is that also your understanding (and the reason this extra synchronization is required)?

Reply:
Actually, the case is more complicated what we see...
1) USM: ideal solution  - via `depends_on(first_event)`  (preferable) or wait() on the first event (future).
2) hetero iterators - you are right  - we have implicit SYCL data access dependency (data dependency graph) and we don't need explicit synch
3) host iterators: due to the first pattern contains a temporary `sycl::buffer `in its own scope, we cannot rely on SYCL data dependency graph. On the second pattern call we have got another temporary `sycl::buffer` and SYCL Runtime will not be aware that it is the same data, I think.

Reply:
That makes sense.  For us to rely on implicit dependencies, we need only a single call which is reused to `__get_sycl_range`.  However, this is something that should be important anyway to avoid extra copies to and from the device with host data.  

Perhaps that rewrite does not belong in this PR, but it seems that may be something we need to audit.  If we are using multiple high level patterns with host iterators, we may be copying data to and from the device for each pattern unnecessarily.